src/libraries/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.MessageMode.Windows.cs

```diff
@@ -0,0 +1,106 @@
+            using NamedPipeClientStream client = CreateClientStream(pipeName, options);
+
+            Task.WaitAll(server.WaitForConnectionAsync(), client.ConnectAsync());
+            client.ReadMode = PipeTransmissionMode.Message;
+
+            ValueTask serverWrite = server.WriteAsync(new byte[] { 1, 2, 3, 4, 5 });
+
+            byte[] buffer1 = new byte[2], buffer2 = new byte[2], buffer3 = new byte[2];
+            bool[] messageCompleted = new bool[3];
+
+            int bytesRead = client.Read(buffer1, 0, 2);
+            messageCompleted[0] = client.IsMessageComplete;
+
+            bytesRead += client.Read(buffer2, 0, 2);
+            messageCompleted[1] = client.IsMessageComplete;
+
+            bytesRead += client.Read(buffer3, 0, 2);
+            messageCompleted[2] = client.IsMessageComplete;
+
+            Assert.Equal(5, bytesRead);
+            Assert.Equal(new byte[] { 1, 2, 3, 4, 5, 0 }, buffer1.Concat(buffer2).Concat(buffer3));
+            Assert.Equal(new bool[] { false, false, true }, messageCompleted);
+
+            await serverWrite;
+        }
+
+        [Theory]
+        [InlineData(PipeTransmissionMode.Byte, PipeOptions.None)]
+        [InlineData(PipeTransmissionMode.Message, PipeOptions.None)]
+        [InlineData(PipeTransmissionMode.Byte, PipeOptions.Asynchronous)]
+        [InlineData(PipeTransmissionMode.Message, PipeOptions.Asynchronous)]
+        public void ServerIn_ClientConnect_Throws(PipeTransmissionMode serverMode, PipeOptions options)
+        {
+            string pipeName = PipeStreamConformanceTests.GetUniquePipeName();
+
+            using NamedPipeServerStream server = new NamedPipeServerStream(pipeName, PipeDirection.In, 1, serverMode, options);
+            using NamedPipeClientStream client = CreateClientStream(pipeName, options);
+
+            Assert.Throws<UnauthorizedAccessException>(() => client.Connect());
+        }
+
+        [Theory]
+        [InlineData(PipeDirection.Out, PipeTransmissionMode.Byte, PipeOptions.None)]
+        [InlineData(PipeDirection.Out, PipeTransmissionMode.Byte, PipeOptions.Asynchronous)]
+        [InlineData(PipeDirection.InOut, PipeTransmissionMode.Byte, PipeOptions.None)]
+        [InlineData(PipeDirection.InOut, PipeTransmissionMode.Byte, PipeOptions.Asynchronous)]
+        public async Task ServerByteMode_ClientReadModeMessage_Throws(PipeDirection serverDirection, PipeTransmissionMode serverMode, PipeOptions options)
+        {
+            string pipeName = PipeStreamConformanceTests.GetUniquePipeName();
+
+            using NamedPipeServerStream server = new NamedPipeServerStream(pipeName, serverDirection, 1, serverMode, options);
+            using NamedPipeClientStream client = CreateClientStream(pipeName, options);
+
+            Task serverConnected = server.WaitForConnectionAsync();
+            client.Connect();
+            await serverConnected;
```

Comment:
Similar question here. I think the `WaitForConnectionAsync` method invocation has to be awaited here, not in line 85.

Additionally, if the server async methods are being called, I would also call the client async methods. In line 84, I would call `await client.ConnectAsync();`

Reply:
We need to have _at least one_ end of the pipe connect asynchronously so that execution isn't blocked and the other side of the connection can be made. As soon as both sides connect, the awaited tasks are completed. The approach here lets me establish that connection and proceed.

I did use `Task.WaitAll(server.WaitForConnectionAsync(), client.ConnectAsync())` elsewhere though, so I can standardize on that so it's clearer.